presto-main/src/main/java/com/facebook/presto/operator/HashBuilderOperator.java

```diff
@@ -550,23 +557,38 @@ private void finishLookupSourceUnspilling()
     {
         checkState(state == State.INPUT_UNSPILLING);
         if (!unspillInProgress.get().isDone()) {
-            // Pages have not be unspilled yet.
+            // Pages have not been unspilled yet.
             return;
         }
-
-        // Use Queue so that Pages already consumed by Index are not retained by us.
-        Queue<Page> pages = new ArrayDeque<>(getDone(unspillInProgress.get()));
-        long memoryRetainedByRemainingPages = pages.stream()
+        List<Page> unspilledPages = getDone(unspillInProgress.get());
+        long sizeOfUnSpilledPages = unspilledPages.stream()
+                .mapToLong(Page::getSizeInBytes)
+                .sum();
+        long retainedSizeOfUnSpilledPages = unspilledPages.stream()
                 .mapToLong(Page::getRetainedSizeInBytes)
                 .sum();
-        localUserMemoryContext.setBytes(memoryRetainedByRemainingPages + index.getEstimatedSize().toBytes(), enforceBroadcastMemoryLimit);
-
-        while (!pages.isEmpty()) {
-            Page next = pages.remove();
-            index.addPage(next);
-            // There is no attempt to compact index, since unspilled pages are unlikely to have blocks with retained size > logical size.
-            memoryRetainedByRemainingPages -= next.getRetainedSizeInBytes();
-            localUserMemoryContext.setBytes(memoryRetainedByRemainingPages + index.getEstimatedSize().toBytes(), enforceBroadcastMemoryLimit);
+        log.debug(
+                "Unspilling for operator %s, unspilled partition %d, sizeOfUnSpilledPages %s, retainedSizeOfUnSpilledPages %s",
+                operatorContext,
+                partitionIndex,
+                succinctBytes(sizeOfUnSpilledPages),
+                succinctBytes(retainedSizeOfUnSpilledPages));
+        if (isSpillInSmallMemoryPool(operatorContext.getSession())) {
+            localUserMemoryContext.setBytes(retainedSizeOfUnSpilledPages + index.getEstimatedSize().toBytes(), enforceBroadcastMemoryLimit);
+            unspilledPages.forEach(index::addPage);
```

Comment:
@rschlussel Because if `unspillInProgress` is not cleared, removing page from queue won't really release memory. As @arhimondr pointed out this function is atomic, it should be safe to clear it and update memory while removing page from queue.

Reply:
I think we might be saying the same thing here.  I'm suggesting nulling out unspillInProgress as soon as we create the queue and leaving everything else as it was before this change.  That way removing the page form the queue will release the memory.